pkg/storage/replica_proposal.go

```diff
@@ -413,7 +415,23 @@ func addSSTablePreApply(
 		// tell Rocks that it is not allowed to modify the file, in which case it
 		// will return and error if it would have tried to do so, at which point we
 		// can fall back to writing a new copy for Rocks to ingest.
-		if _, err := os.Stat(path); err == nil {
+		if stat, err := os.Stat(path); err == nil {
+			canIngestRaftFile = true
+			// HACK: RocksDB does not like ingesting the same file (by inode) twice.
+			// If we ingested this file before it will have links from raft and rocks,
+			// so we should not allow ingesting the raft copy again. If we were _sure_
+			// that the ingest succeeded we could potentially no-op, but it is safer
+			// to assume it didn't and reingest
+			if sys := stat.Sys(); sys != nil {
+				if s, ok := sys.(*syscall.Stat_t); ok && s.Nlink > 1 {
```

Comment:
I wonder if it's safer to do copies when `!ok || s.Nlink > 1`. Looks like on windows and maybe other platforms the cast will fail.

Reply:
I think on windows we won't be able to link to it anyway?

Reply:
RocksDB looks like it's able to create a hardlink on Windows. Were you referring to that or the link made in Cockroach before calling RocksDB's file ingestion?

Reply:
ah, fair enough. Inverted it.